### PR TITLE
Reapply [IR] Add new Range attribute using new ConstantRange Attribute type

### PR DESCRIPTION
**Pull request by andjo403**
only change from https://github.com/llvm/llvm-project/pull/83171 it the change of the allocator so the destructor is called for ConstantRangeAttributeImpl

reverts https://github.com/llvm/llvm-project/pull/84549

CC @nikic

### CODE REVIEW

File: llvm/lib/IR/LLVMContextImpl.h [hunk 1]
```diff
@@ -14,6 +14,7 @@
 #ifndef LLVM_LIB_IR_LLVMCONTEXTIMPL_H
 #define LLVM_LIB_IR_LLVMCONTEXTIMPL_H
 
+#include "AttributeImpl.h"
```
Comment (nikic) [inline llvm/lib/IR/LLVMContextImpl.h:17]: Would a forward-declaration for ConstantRangeAttributeImpl be sufficient here?

Comment (andjo403) [reply llvm/lib/IR/LLVMContextImpl.h:17]: it compiles add check-llvm passes with forward-declaration also so seems like it :D

**Comment by andjo403** [comment]
I can not find how to run with the sanitizer to verify that the Buildbots do not fail.

**Comment by dtcxzyw** [comment]
> I can not find how to run with the sanitizer to verify that the Buildbots do not fail.

You can pass `-DLLVM_USE_SANITIZER=Memory` to cmake.

**Comment by andjo403** [comment]
hmm seems like I have a too old default compiler 
`c++: error: unrecognized argument to ‘-fsanitize=’ option: ‘memory’`

**Comment by nikic** [comment]
The right sanitizer for this is `-DLLVM_USE_SANITIZER=Address`. Though you could also just run the `Bitcode/attributes.ll` test though valgrind to check for leaks.

**Comment by andjo403** [comment]
thanks @nikic the Address sanitizer worked, was able to reproduce the failure and that the failure was gone with this PR so if nothing else think this PR is ready

**Review by nikic — APPROVED** [review]
LGTM

**Comment by andjo403** [comment]
I need help with merge I do not have rights to do it

**Pull request merged**